ports/raspberrypi/common-hal/displayio/ParallelBus.c

```diff
@@ -32,37 +32,128 @@
 #include "py/runtime.h"
 #include "shared-bindings/digitalio/DigitalInOut.h"
 #include "shared-bindings/microcontroller/__init__.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/rp2pio/StateMachine.h"
+
+static const uint16_t parallel_program[] = {
+// .side_set 1
+// .wrap_target
+    0x6008, // out pins, 8 side 0
+    0xB042  // nop         side 1
+// .wrap
+};
 
 void common_hal_displayio_parallelbus_construct(displayio_parallelbus_obj_t* self,
     const mcu_pin_obj_t* data0, const mcu_pin_obj_t* command, const mcu_pin_obj_t* chip_select,
-    const mcu_pin_obj_t* write, const mcu_pin_obj_t* read, const mcu_pin_obj_t* reset) {
+    const mcu_pin_obj_t* write, const mcu_pin_obj_t* read, const mcu_pin_obj_t* reset, uint32_t frequency) {
+
+    uint8_t data_pin = data0->number;
+    for (uint8_t i = 0; i < 8; i++) {
+        if (!pin_number_is_free(data_pin + i)) {
+            mp_raise_ValueError_varg(translate("Bus pin %d is already in use"), i);
+        }
+    }
+
+    uint8_t write_pin = write->number;
+    if (!pin_number_is_free(write_pin)) {
+        mp_raise_ValueError_varg(translate("Bus pin %d is already in use"), write_pin);
+    }
+
+    self->command.base.type = &digitalio_digitalinout_type;
+    common_hal_digitalio_digitalinout_construct(&self->command, command);
+    common_hal_digitalio_digitalinout_switch_to_output(&self->command, true, DRIVE_MODE_PUSH_PULL);
+
+    self->chip_select.base.type = &digitalio_digitalinout_type;
+    common_hal_digitalio_digitalinout_construct(&self->chip_select, chip_select);
+    common_hal_digitalio_digitalinout_switch_to_output(&self->chip_select, true, DRIVE_MODE_PUSH_PULL);
+
+    self->read.base.type = &digitalio_digitalinout_type;
+    common_hal_digitalio_digitalinout_construct(&self->read, read);
+    common_hal_digitalio_digitalinout_switch_to_output(&self->read, true, DRIVE_MODE_PUSH_PULL);
+
+    self->data0_pin = data_pin;
+    self->write = write_pin;
+
+    self->reset.base.type = &mp_type_NoneType;
+    if (reset != NULL) {
+        self->reset.base.type = &digitalio_digitalinout_type;
+        common_hal_digitalio_digitalinout_construct(&self->reset, reset);
+        common_hal_digitalio_digitalinout_switch_to_output(&self->reset, true, DRIVE_MODE_PUSH_PULL);
+        never_reset_pin_number(reset->number);
+        common_hal_displayio_parallelbus_reset(self);
+    }
 
-    mp_raise_NotImplementedError(translate("ParallelBus not yet supported"));
-    // TODO: Implement with PIO and DMA.
+    never_reset_pin_number(command->number);
+    never_reset_pin_number(chip_select->number);
+    never_reset_pin_number(write_pin);
+    never_reset_pin_number(read->number);
+    for (uint8_t i = 0; i < 8; i++) {
+        never_reset_pin_number(data_pin + i);
+    }
+
+    common_hal_rp2pio_statemachine_construct(&self->state_machine,
+        parallel_program, sizeof(parallel_program) / sizeof(parallel_program[0]),
+        frequency, // frequency
```

Comment:
I think you may actually want:
```suggestion
        frequency * 2, // frequency
```

This accounts for it taking two instructions per clock cycle output.

Reply:
I tried doubling it and it doesn't work. To reflect the true frequency probably best to add the `* 2` in? (vs the frequency of the PIO program which is the current raw number).

Reply:
What does it do? Ya, the ParallelBus frequency should be the frequency of the output clock signal, not the underlying state machine.

Reply:
The display starts tearing/ripping. I wasn't able to find it in the spec sheet for the driver chip but I'm guessing hitting the max it can take.

I'll change the frequency to take the real frequency and multiply it by 2 to give to the statemachine.

Reply:
Ya, I think you should lower the given frequency then, not change the math.

Reply:
Note that tearing could also happen if you are updating the display faster than it is refreshing too.

Reply:
Made it so frequency is the actual frequency and took the PIO instruction size into account. Then changed the default do to 30Mhz which works fine. Any higher and weird artifacts appear or it simply doesn't work.